### DISCUSSION TITLE
getting completions from Nek5000 although not installed on my system

### DISCUSSION THREAD

**Discussion by zbinkz** [fortran-lang/fortls#183] [General]
Hi, running `fortls` version 2.12.0 installed from pip, within Kate 20.12.2 on Debian Bullseye with Plasma 5.20.5. `fortls` starts correctly in Kate LSP client plugin.

My local settings are the following:
```
{
    "servers": {
        "fortran": {
            "command": ["/home/zbinkz/.local/bin/fortls"],
            "rootIndicationFileNames": [".fortls"],
            "url": "https://github.com/gnikit/fortls",
            "highlightingModeRegex": "^Fortran.*$"
        }
    }
}
```
The global settings are still relative to package `fortran-language-server-1.12.0` which I had installed before switching to `fortls-2.12` from pip, but local settings are supposed to supersede them.

What is weird is that I'm getting completions from Nek5000 although it's not installed on my system. No idea why it's happening and this is the first time I try `fortls`. Has anybody seen that before?

Thanks

**Comment by gnikit** [comment]
A couple of questions:

1. What are your local/global fortls settings?

2. What type of completions are you getting?

To share your completions you can use the following debug options from a terminal https://gnikit.github.io/fortls/options.html#debug-options-command-line-only

You will have to do something like this

```sh
fortls --debug_rootpath ./your/projects root --debug_filepath ./your/filename --debug_completion --debug_line yourlineNumber --debug_char yourColumnNumber
```

**Comment by zbinkz** [comment]
1. What are your local/global fortls settings?

Local settings:
```
{
    "servers": {
        "fortran": {
            "command": ["/home/zbinkz/.local/bin/fortls"],
            "rootIndicationFileNames": [".fortls"],
            "url": "https://github.com/gnikit/fortls",
            "highlightingModeRegex": "^Fortran.*$"
        }
    }
}
```
Global settings:
```
"fortran": {
            "command": ["fortls"],
            "rootIndicationFileNames": [".fortls"],
            "url": "https://github.com/hansec/fortran-language-server",
            "highlightingModeRegex": "^Fortran.*$"
        }
```

2. What type of completions are you getting?
```
fortls --debug_rootpath . --debug_filepath navierStokesFV.f90 --debug_completion --debug_line 21 --debug_char 4

Testing "initialize" request:
  Root = "."
[INFO - 11:59:51] fortls - Fortran Language Server 2.12.0 Initialized
  Successful!

  Source directories:
    /home/zbinkz/HGST/Tutorials_and_Courses/MPI/test_fvm_mpi/navierStokesFV

Testing "textDocument/completion" request:
  File = "navierStokesFV.f90"
  Line = 21
  Char = 4

  Results:
    3: mesh -> SUBROUTINE mesh(NX, NY, SIDEX, SIDEY, dx, dy, vol, x, y)
```
The completion above, although correct, is not the whole story. As I type "mesh" I get other completion suggestions unrelated to the current project. Here they are typed below:
```
FUNCTION MESH_IS_3D(newval)
SUBROUTINE Meshconv(c_input_basename, input_basename_len, c_input_mesh_format, input_mesh_format_len, c_output_mesh_format, output_mesh_format_len) !! 
```

The completion suggestions I used to get from Nek5000 turned out to be from an old NekRS folder I had on my drive, even though it was not installed and not in my $PATH. The same for WRF completions. I compressed both folders and deleted the source and now the completion suggestions are gone.

**Reply by gnikit** [reply]
It's not because things are in your path. By default `fortls` parses all files under the root directory and it will give suggestions based on those. You should have a read at the configuration options for the server (https://gnikit.github.io/fortls/options.html) you can include/exclude directories and file extensions per project or globally.

If you want to have more than 1 root folders you would have to check if `emacs-lsp` supports that. In our VS Code extension for example `fortls` can be set up with multiple workspace Folders (i.e. root directories).

**Reply by zbinkz** [reply]
I did read these parts of the configuration options before setting `fortls` up, and figured it would parse within my project root dir. The WRF and NekRS folders were on the same drive in completely different folders though. They could not be reached recursively from within my project dir unless `fortls` scans the entire drive. Is that its intended behaviour?

If I use a project configuration file with `source_dirs` and/or `include_dirs` specified, will `fortls` parse exclusively in these locations?

**Reply by gnikit** [reply]
I think your global configuration options might be taking precedence over the local ones. Try and remove the global LSP options for Fortran. The global options will parse your entire home, so not very helpful IMO.

Another explanation of this strange behaviour is if `emacs-lsp` spawns 2 instances of `fortls`, one with local and one with global settings. You can check in your active processes, there should be only 1 `fortls` instance running.

 >If I use a project configuration file with source_dirs and/or include_dirs specified, will fortls parse exclusively in these locations?

Yes, you can also use configuration files (JSON format) instead of command line arguments, makes the process of configuring `fortls` substantially easier. Add a file by the name `.fortls` in your root dir and `fortls` will read those options. If both command line options and config file options have been provided the config file options take precedence.

**Reply by zbinkz** [reply]
Okay, using a project-local `.fortls` file and completions are strictly from the project now.

Filed a bug report with `kde.org` anyway cause the editor's behaviour is definitely buggy. The old settings should not persist.

Thanks for the help @gnikit!

**Reply by gnikit** [reply]
Glad you were able to resolve this
